Add "reference" help topic

rista404 opened the pull request:
Hello! I tried to keep the reference simple and in line with [the manual](https://cli.github.com/manual/). Let me know if I need to change anything!

<img width="1095" alt="image" src="https://user-images.githubusercontent.com/8061997/96385447-4c6d5e00-1194-11eb-8422-188a58956651.png">

Closes #1734

rista404 committed:
Add "reference" help topic

mislav reviewed (commented):
This looks really cool! Hvala @rista404 for working on this 💖

vilmibm reviewed (dismissed):
This is really cool! Thank you!

Please update this so it is accessed via `gh help reference` instead of having it print as part of `gh help` or as usage information. It's too much information to print in those cases but is a great help topic.

rista404 committed:
Only print reference as a help topic

rista404 commented:
@vilmibm totally missed that it is printed in `gh help` also! Fixed it.

vilmibm commented:
We just merged some stuff that breaks this; I'm going to push some commits to fix it so it can be merged.

Commit:
Merge remote-tracking branch 'origin/trunk' into cmd-reference-1734

Commit:
fix for color fns, slightly generalize

vilmibm commented:
for posterity: there was an internal concern about generating this on every invocation of `gh`; I ran a ton of benchmarks on my machine using a non-networked command and there was no meaningful effect on performance from having this in.

Okeakwalam27 commented:
Nice

On Wed, Nov 4, 2020 at 10:13 PM Nate Smith <notifications@github.com> wrote:

>
>
> for posterity: there was an internal concern about generating this on
> every invocation of gh; I ran a ton of benchmarks on my machine using a
> non-networked command and there was no meaningful effect on performance
> from having this in.
>
>
>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cli/cli/pull/2223#issuecomment-721978099>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQZUFQF2IL53VW67UPA57C3SOG7XTANCNFSM4SVKTDSQ>
> .
>
>
>

vilmibm commented:
While I'm still in favor of merging this as it's very useful from the command line (`gh help reference`) it unfortunately does not survive the translation to markdown well for use on the manual site; it ends up being very garbled with all formatting and indentation lost.

I see three options:

- porting this to generate markdown (instead of directly using `ColorScheme` functions) and then running it through glamour
- porting it to generate markdown and printing the raw markdown on the CLI
- tweaking the formatting so it looks at least intelligible when it becomes markdown

I'm leaning towards the first option; in general the help topic system is meant to do double duty of CLI and HTML help and we're only going to want _more_ markdown formatting in help topics, not less.

I'm happy to augment this to output markdown.

Commit:
WIP for switching to markdown

vilmibm commented:
I've gotten this generating markdown in my local environment, but some shortcomings of glamour are making the output less than ideal; the primary issue is the inability to manually linebreak which I'd like for the flag usage lists. I'll explore using tables or blockquotes for that.

Either way I have to focus on other things today but I'll circle back to this.

vilmibm committed:
escape gt/lt

vilmibm committed:
minor

vilmibm committed:
higher wrap point

vilmibm committed:
detect terminal theme

vilmibm committed:
futz with angle brackets once more

vilmibm committed:
minor cleanup

vilmibm commented:
I've pushed commits for having this output markdown. I had to make some trade-offs but I think it looks acceptable; most importantly, it should look ok on the manual website.

![image](https://user-images.githubusercontent.com/98482/99325002-f47d5200-283a-11eb-8cfb-aaf60f1f2605.png)

vilmibm committed:
prepend parent commands

samcoe commented on pkg/cmd/root/help_reference.go:
```diff
@@ -0,0 +1,82 @@
+func cmdRef(cs *iostreams.ColorScheme, cmd *cobra.Command, parent string) string {
+	ref := ""
+
+	if cmd.Hidden {
```
As this is a reference for all of `gh` I am wondering if we should include the `hidden` commands also?

billygriffin replied:
I think my preference is no. If we choose to hide a command for some reason, that's intentional and therefore isn't intended to be part of the documentation. Definitely open to changing my mind on that, but part of the fun of having a few easter eggs is finding them.

samcoe commented on pkg/cmd/root/help_topic.go:
```diff
@@ -47,6 +47,26 @@ var HelpTopics = map[string]map[string]string{
 	},
 }
 
+func NewHelpTopicFn(topic, short string, fn func() string) *cobra.Command {
```
What do you think about changing this name to `NewDynamicHelpTopic` and changing `NewHelpTopic` to `NewStaticHelpTopic`? My thoughts are that they are a bit more descriptive, and help make the distinction between the two a bit better.

vilmibm replied:
good idea

samcoe commented on pkg/cmd/root/help_reference.go:
```diff
@@ -0,0 +1,82 @@
+package root
```
Now that we have more than one help topic, does it make sense to move the help and help topic code into their own package?

vilmibm replied:
i'm ok with kicking that can a little further down the road.

vilmibm committed:
rename help topic fns and add test

mislav committed:
Simplify reference help generation

- the `<...>` characters from command usage line are now preserved by enclosing the entire usage synopsis in a code span
- hard breaks in flag usage lines are preserved by enclosing flag usage in a code block
- TTY detection and Markdown rendering are now delayed until the user explicitly requests `gh help reference`
- `gh help reference` output is now pager-enabled

mislav reviewed (approved):
Great work everyone!

I've pushed some changes to simplify things and add PAGER support. This is how `gh reference` looks now:

<img width="810" alt="Screen Shot 2020-11-17 at 20 57 36" src="https://user-images.githubusercontent.com/887/99442458-b440d100-2919-11eb-8ff1-506e044fd64a.png">

I've avoided Markdown issues around `<...>` and line breaks by wrapping those sections in `<code>` spans or blocks. Particularly, `FlagUsages()` output should be rendered in a code block due to its output currently being vertically-aligned.

When stdout is not a terminal, the `reference` command outputs no color. Similarly, no Markdown rendering or terminal detection happens when initializing the `reference` command, since that would add to the overhead of executing any `gh` command. Instead, rendering and detecting terminal features now only happens if the reference output was actually requested.

@vilmibm @samcoe When rebasing my changes onto this branch, I've removed the `Static/DynamicHelpTopic` distinction in the cleanup. It didn't feel that right now we need to maintain that distinction, since each of the functions had precisely 1 caller. It was also relatively easy to implement this with our original `NewHelpTopic()` function without any modifications.

vilmibm commented on pkg/cmd/root/root.go:
```diff
@@ -92,9 +92,14 @@ func NewCmdRoot(f *cmdutil.Factory, version, buildDate string) *cobra.Command {
 	cmdutil.DisableAuthCheck(cmd)
 
+	// this needs to appear last:
+	referenceCmd.Long = referenceLong(cmd)
```
I dislike this approach but don't see value in further back and forthing on this given that we only have two help topics

vilmibm commented on pkg/cmd/root/help_reference.go:
```diff
@@ -0,0 +1,69 @@
+
+		_ = io.StartPager()
+		defer io.StopPager()
+		fmt.Fprint(io.Out, md)
```
good thinking.

the environment help topic is also long and i assume future help topics will also be greater than one "page" (which seems to be the point of them) so we might want to consider always running the pager for help topics in the future

vilmibm reviewed (approved):
I didn't think of nesting `code`s into `h2`s, i'm glad that works.

I had considered doing the `pre` approach for flags but was worried about over indentation; it looks fine though :+1:
